### PR TITLE
Allow tuples with a single element

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
See https://github.com/p4lang/p4c/pull/2518 and https://github.com/p4lang/p4c/pull/2514

### CODE REVIEW

**Comment by jafingerhut** [comment]
The grammar shows that a keySetExpression is basically a comma-separated list of one or more simpleKeySetExpressions.

tupleKeySetExpression has parentheses around it.
simpleExpression list does not have parentheses around it.

It seems somewhat odd that parentheses are optional?  Maybe p4c checks in certain contexts that they must be present, and must be absent in others?  I haven't dug into that very far.

**Comment by mbudiu-vmw** [comment]
I don't remember why we haven't merged this one.

File: p4-16/spec/grammar.mdk [hunk 1]
```diff
@@ -178,14 +178,22 @@ keysetExpression
     ;
 
 tupleKeysetExpression
-    : '(' simpleKeysetExpression ',' simpleExpressionList ')'
+    : "(" simpleKeysetExpression "," simpleExpressionList ")"
+    | "(" reducedSimpleKeysetExpression ")"
```
Comment (jfingerh) [inline p4-16/spec/grammar.mdk:182]: Why not use simpleKeysetExpression here?  What is the need for introducing reducedSimpleKeysetExpression?

File: p4-16/spec/grammar.mdk [hunk 2]
```diff
@@ -178,14 +178,22 @@ keysetExpression
...
     | simpleExpressionList ',' simpleKeysetExpression
     ;
 
+reducedSimpleKeysetExpression
+    : expression "&&&" expression
+    | expression ".." expression
+    | DEFAULT
+    | "_"
+    ;
+
```
Comment (jfingerh) [inline p4-16/spec/grammar.mdk:196]: If there is a reason to introduce reducedSimpleKeysetExpression, it seems reasonable for humans reviewing and understanding this to make its definition as similar to simpleKeysetExpression as possible, e.g. something like "DEFAULT | DONTCARE | expression MASK expression | expression RANGE expression"

**Comment by jnfoster** [comment]
We discussed at the March 1st, 2021 LDWG meeting. 

@jafingerhut agreed to shepherd it and @mbudiu-vmw will check consistency with the `p4c` implementation.

**Comment by mbudiu-vmw** [comment]
I can confirm that this grammar matches the compiler; there is some duplication for avoiding conflicts.

File: p4-16/spec/grammar.mdk (hunk 1 above)
Comment (mbudiu-vmw) [reply p4-16/spec/grammar.mdk:182]: reducedKeysetExpression is missing a case, `expression`

**Comment by jfingerh** [comment]
@mbudiu-vmw I have carefully compared this modified grammar with the latest p4c, and agree that it matches in regards to the changes here.  I will resolve all of my comments/questions on this PR, and approve it for merging.

**Comment by jfingerh** [comment]
@mbudiu-vmw Rather, I thought I was going to resolve my comments, but I don't see the button for Github to allow me to do that.  I have earlier approved this PR, so if there is anything else you would like me to do to approve this PR, let me know.

**Comment by mbudiu-vmw** [comment]
Do you have 2 distinct github accounts?
I think we can just merge it.

**Comment by jfingerh** [comment]
I do, @jfingerh and @jafingerhut, which I unfortunately confuse myself with, as well as others.  Sorry about that.

**Pull request merged**